Libraries/LibMedia/Audio/PlaybackStreamWasapi.cpp

```diff
@@ -0,0 +1,508 @@
+{
+    HRESULT hr;
+    if (!s_com_uninitializer.initialized) {
+        TRY_HR(CoInitializeEx(NULL, COINIT_MULTITHREADED));
+        s_com_uninitializer.initialized = true;
+    }
+
+    auto state = make_ref_counted<PlaybackStreamWASAPI::AudioState>();
+
+    TRY_HR(CoCreateInstance(__uuidof(MMDeviceEnumerator), NULL, CLSCTX_ALL, IID_PPV_ARGS(&state->enumerator)));
+    TRY_HR(state->enumerator->GetDefaultAudioEndpoint(eRender, eConsole, &state->device));
+    TRY_HR(state->device->Activate(__uuidof(IAudioClient), CLSCTX_ALL, nullptr, &state->audio_client));
+
+    state->data_request_callback = move(data_request_callback);
+
+    WAVEFORMATEXTENSIBLE* device_format;
+    state->audio_client->GetMixFormat(reinterpret_cast<WAVEFORMATEX**>(&device_format));
+    ScopeGuard free_mix_format = [&device_format] { CoTaskMemFree(device_format); };
+
+    dbgln("Mixing engine audio format:\n");
+    print_audio_format(device_format);
+
+    Vector<Channel, 32> channel_vec;
+    VERIFY(device_format->Format.nChannels <= 32);
+    VERIFY(device_format->dwChannelMask != 0);
+    auto channels = device_format->Format.nChannels;
+
+    // Conveniently the enum is in the same order. The difference is that KSMedia uses a bitmask instaead
+    for (u64 i = 0; i < channels; i++) {
+        u32 mask = 1 << i;
+        if (device_format->dwChannelMask & mask)
+            // i + 1 as the first bit is at i == 0
+            channel_vec.append(static_cast<Channel>(i + 1));
+    }
+    ChannelMap channel_map { channel_vec };
+
+    auto sample_specification = SampleSpecification { device_format->Format.nSamplesPerSec, channel_map };
+    sample_specification_callback(sample_specification);
+
+    // Set up whatever we were given.
+
+    state->wave_format->Format.wFormatTag = WAVE_FORMAT_EXTENSIBLE;
+    state->wave_format->Format.nChannels = device_format->Format.nChannels;
+    state->wave_format->Format.nSamplesPerSec = device_format->Format.nSamplesPerSec;
+    state->wave_format->Format.nAvgBytesPerSec = device_format->Format.nAvgBytesPerSec;
+    state->wave_format->Format.nBlockAlign = device_format->Format.nBlockAlign;
+    state->wave_format->Format.wBitsPerSample = device_format->Format.wBitsPerSample;
+    state->wave_format->Format.cbSize = sizeof(WAVEFORMATEXTENSIBLE) - sizeof(WAVEFORMATEX);
+    state->wave_format->Samples.wValidBitsPerSample = device_format->Samples.wValidBitsPerSample;
+    state->wave_format->dwChannelMask = device_format->dwChannelMask;
+    state->wave_format->SubFormat = device_format->SubFormat;
+
+    // FIXME: From what I understand the audio mixer uses 32bit floats internally. If that is not always the case we need to handle it.
+    VERIFY(device_format->SubFormat == KSDATAFORMAT_SUBTYPE_IEEE_FLOAT);
+    VERIFY(device_format->Format.wBitsPerSample == 32);
```

Comment:
Instead of verifying this, could we just create a format that reuses the sample rate and channel count/mask, but with the float subformat? Then we don't need a verify here, and we can just return an error if the API rejects the format.